zkevm-circuits/src/util/word.rs

```diff
@@ -100,6 +117,90 @@ impl<F: Field, const N: usize> WordLimbs<Cell<F>, N> {
         })
     }
 
+    /// assign bytes to wordlimbs first half/second half respectively
+    // N_LO, N_HI are number of bytes to assign to first half and second half of size N limbs,
+    // respectively N_LO and N_HI can be different size, the only requirement is N_LO % (N/2)
+    // and N_HI % (N/2) [N/2] limbs will be assigned separately.
+    // E.g. N_LO = 4 => [nl1, nl2, nl3, nl4]
+    // N_HI = 2 => [nh1, nh2]
+    // N = 2 => [l1, l2]
+    // it equivalent l1.assign(nl1.expr() + nl2.expr() * 256 + nl3.expr() * 256^2 +  nl3.expr() *
+    // 256^3) and l2.assign(nh1.expr() + nh2.expr() * 256)
+    pub fn assign_lo_hi<const N_LO: usize, const N_HI: usize>(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        bytes_lo: Option<[u8; N_LO]>,
+        bytes_hi: Option<[u8; N_HI]>,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error> {
+        assert_eq!(N % 2, 0); // TODO use static_assertion instead
+        assert_eq!(N_LO % (N / 2), 0);
+        assert_eq!(N_HI % (N / 2), 0);
+        // assign lo
+        let bytes_lo_assigned: Result<Vec<AssignedCell<F, F>>, Error> =
+            bytes_lo.map_or(Err(Error::Synthesis), |bytes| {
+                bytes
+                    .chunks(N_LO / (N / 2)) // chunk in little endian
+                    .map(|chunk| from_bytes::value(chunk))
+                    .zip(self.limbs[0..(N / 2)].iter())
+                    .map(|(value, cell)| cell.assign(region, offset, Value::known(value)))
+                    .collect()
+            });
+
+        // assign hi
+        let bytes_hi_assigned: Result<Vec<AssignedCell<F, F>>, Error> =
+            bytes_hi.map_or(Err(Error::Synthesis), |bytes| {
+                bytes
+                    .chunks(N_HI / (N / 2)) // chunk in little endian
+                    .map(|chunk| from_bytes::value(chunk))
+                    .zip(self.limbs[N / 2..].iter())
+                    .map(|(value, cell)| cell.assign(region, offset, Value::known(value)))
+                    .collect()
+            });
+
+        match (bytes_lo_assigned, bytes_hi_assigned) {
+            (Ok(bytes_lo_assignedcells), Ok(bytes_hi_assignedcells)) => Ok([
+                bytes_lo_assignedcells.to_vec(),
+                bytes_hi_assignedcells.to_vec(),
+            ]
+            .concat()),
+            _ => Err(Error::Synthesis),
+        }
+    }
+
+    /// assign u256 to wordlimbs
+    pub fn assign_u256(
```

Comment:
Does this function work for only ` WordLimbs<Cell<F>, 2>` or does it work for  `WordLimbs<Cell<F>, 4>`, `WordLimbs<Cell<F>, 8>`, `WordLimbs<Cell<F>, 16>`, and `WordLimbs<Cell<F>, 32>`.

We only have at most 5 cases, but how we do it seems very involving.

Reply:
it works for all `WordLimbs<Cell<F>, N>` and inner function `assign_lo_hi` will help to handle all the case